retrieval/scrape.go

```diff
@@ -434,7 +443,8 @@ func newScrapeLoop(ctx context.Context, sc scraper, app, reportApp func() storag
 		scraper:        sc,
 		appender:       app,
 		reportAppender: reportApp,
-		cache:          map[string]uint64{},
+		refCache:       map[string]uint64{},
+		lsetCache:      map[uint64]lsetCacheEntry{},
```

Comment:
Yeah, I noticed that. I've it down as future work, the tsdb API is going to change too which will require a rejig here. TODO added.

> Actually maybe the same staleness logic can be reused to purge stale cache entries?

More simply you can keep just the entries from the last scrape.